src/navigation/graph_domain.h

```diff
@@ -232,39 +268,62 @@ struct GraphDomain {
     e.max_clearance = max_clearance;
     e.edge.p0 = states[s0].loc;
     e.edge.p1 = states[s1].loc;
+    e.has_door = has_door;
+    e.has_stairs = has_stairs;
+    edges.push_back(e);
+  }
+
+  void AddUndirectedEdge(const json& edge) {
+    uint64_t s0 = edge["s0_id"].get<uint64_t>();
+    uint64_t s1 = edge["s1_id"].get<uint64_t>();
+    CHECK_LT(s0, states.size());
+    CHECK_LT(s1, states.size());
+    edges.push_back(NavigationEdge::fromJSON(edge, states));
+  }
+
+  // This should only be called while upgrading v1 maps
+  void AddUndirectedEdge(const uint64_t s0, const uint64_t s1) {
+    GrowIfNeeded(s0);
+    GrowIfNeeded(s1);
+    NavigationEdge e;
+    e.s0_id = s0;
+    e.s1_id = s1;
+    e.edge.p0 = states[s0].loc;
+    e.edge.p1 = states[s1].loc;
+    e.has_door = false;
+    e.has_stairs = false;
     edges.push_back(e);
   }
 
   uint64_t AddDynamicState(const Eigen::Vector2f& v) {
     static const bool kDebug = false;
     CHECK(!states.empty());
     // Find the closest Edge.
-    const NavigationEdge e = GetClosestEdge(v);
-    const uint64_t p0_id = e.s0_id, p1_id = e.s1_id;
-    const Eigen::Vector2f& p0 = states[p0_id].loc;
-    const Eigen::Vector2f& p1 = states[p1_id].loc;
-    if (kDebug) {
-      printf("Adding %f,%f\n", v.x(), v.y());
-      printf("p0: %lu (%f,%f)\n", p0_id, p0.x(), p0.y());
-      printf("p1: %lu (%f,%f)\n", p1_id, p1.x(), p1.y());
+    NavigationEdge closest_edge;
+    float closest_dist = FLT_MAX;
+    if (!GetClosestEdge(v, closest_edge, &closest_dist)) {
+      std::cerr << "Unable to find closest edge to point " << v << std::endl;
+      return 0;
     }
     // Find the projection of v on to the line segment p0 : p1;
     const Eigen::Vector2f pmid =
-        geometry::ProjectPointOntoLineSegment(v, p0, p1);
+        geometry::ProjectPointOntoLineSegment(v, closest_edge.edge.p0, closest_edge.edge.p1);
 
     // Add p0 : pmid
     const uint64_t pmid_id = AddState(pmid);
-    AddUndirectedEdge(p0_id, pmid_id, e.max_speed, e.max_clearance);
+    const uint64_t p0_id = closest_edge.s0_id;
+    const uint64_t p1_id = closest_edge.s1_id;
+    AddUndirectedEdge(p0_id, pmid_id, closest_edge.max_speed, closest_edge.max_clearance, false, false);
 
     // Add p1 : pmid
-    AddUndirectedEdge(p1_id, pmid_id, e.max_speed, e.max_clearance);
+    AddUndirectedEdge(p1_id, pmid_id, closest_edge.max_speed, closest_edge.max_clearance, false, false);
```

Comment:
This assumes that the correct thing to do is to say that the dynamic edge does not have a door, and it is not on stairs. It should be made explicit with a comment. I don't think we have a use case to distinguish this right now, but we might need to in the future.